### PR TITLE
feat: adding the use of topsort sdk for reporting events

### PR DESCRIPTION
**Pull request by barbmarcio**
This PR adds the use of Topsort.js SDK for reporting events

You can see the use inside analytics in [this](https://www.loom.com/share/6b54e16129fb4037b613418a40ce53c3?sid=0ca1063e-90d6-46d3-8ccc-f3c13b6b8d3a) video

| File | Before (gzip) | After (gzip) |
|:-------|:-------|:-------|
| dist/ts.mjs | 9.04 kB (3.21 kB) | 10.37 kB (3.61 kB) |
| dist/ts.js | 6.32 kB (2.72 kB) |  7.63 kB (3.22 kB) |

The difference comes from:
- The error handling in the SDK ([AppError](https://github.com/Topsort/topsort.js/blob/main/src/lib/app-error.ts))
- [Configuration validation](https://github.com/Topsort/topsort.js/blob/main/src/lib/validate-config.ts) (used to verify required fields [coming](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L40) in the request)
- `withValidation` ([HOF](https://github.com/Topsort/topsort.js/blob/main/src/lib/with-validation.ts#L4) used to wrap functions that we will validate fields)
- [Request handling](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L33) (The class [API Client](https://github.com/Topsort/topsort.js/blob/main/src/lib/api-client.ts#L6) we use to wrap the function calls for better error handling)
- Use of [URL construction](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L26) (used for better error messages for invalid URL) - new URL(url)

Please feel free to check the diff [here](https://gist.github.com/barbmarcio/a05c5d4d1638955d49a42719dde8edb0) as well

### CODE REVIEW

File: src/detector.ts [hunk 1]
```diff
@@ -132,12 +131,19 @@ async function processor(data: ProductEvent[]): Promise<ProcessorResult> {
...
     promises.push(
-      reportEvent(getApiPayload(entry), window.TS)
+      reportEvent(config, getApiPayload(entry))
         .then((result) => {
-          const q = result.retry ? r.retry : r.done;
-          q.add(entry.id);
+          if (result.ok) {
+            r.done.add(entry.id);
+          } else {
+            r.retry.add(entry.id);
```
Comment (sk-) [inline src/detector.ts:145]: Note that not all response code should be retried. Only 429 and 5xx.

File: src/reporter.ts [hunk 2]
```diff
@@ -1,29 +0,0 @@
...
-export async function reportEvent(e: TopsortEvent, config: Config) {
-  try {
-    const url = (config.url || "https://api.topsort.com") + "/v2/events";
-    const r = await fetch(url, {
-      method: "POST",
-      headers: {
-        "Content-Type": "application/json",
-        // Can't use User-Agent header because of
-        // https://bugs.chromium.org/p/chromium/issues/detail?id=571722
-        "X-UA": `ts.js/${version}`,
```
Comment (sk-) [inline src/reporter.ts:18]: Note that we will be loosing the ability to know which version of analytics.js they are using. It'd be great if we could add a parameter that gets added to the user agent sent by topsort sdk.

**Review by sk- — COMMENTED** [review]
Overall looks good. Thanks for the change. My only concern is about the resulting library size and whether they builder is actually applying tree shaking for the dependencies.

Could you add to the PR description the size of the library before and after the change.

File: src/detector.ts (hunk 1 above)
Comment (barbmarcio) [reply src/detector.ts:145]: Will adapt to have it correctly coming from the sdk! Thank you very much

**Comment by jbergstroem** [comment]
I would like to understand why the size grows. Since its so small, can you peek into the diff?

**Comment by barbmarcio** [comment]
> I would like to understand why the size grows. Since its so small, can you peek into the diff?

The difference comes from:
- The error handling in the SDK ([AppError](https://github.com/Topsort/topsort.js/blob/main/src/lib/app-error.ts))
- [Configuration validation](https://github.com/Topsort/topsort.js/blob/main/src/lib/validate-config.ts) (used to verify required fields [coming](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L40) in the request)
- `withValidation` ([HOF](https://github.com/Topsort/topsort.js/blob/main/src/lib/with-validation.ts#L4) used to wrap functions that we will validate fields)
- [Request handling](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L33) (The class [API Client](https://github.com/Topsort/topsort.js/blob/main/src/lib/api-client.ts#L6) we use to wrap the function calls for better error handling)
- Use of [URL construction](https://github.com/Topsort/topsort.js/blob/main/src/functions/events.ts#L26) (used for better error messages for invalid URL) - new URL(url)

Please feel free to check the diff [here](https://gist.github.com/barbmarcio/a05c5d4d1638955d49a42719dde8edb0) as well

**Review by sk- — APPROVED** [review]

**Review by PabloReszczynski — APPROVED** [review]

**Pull request merged**